### PR TITLE
Metric Comparison + API Updates

### PR DESCRIPTION
**Pull request by donn**
* `openlane.common`
  * Created new function `get_latest_file`, that finds the latest file with a certain filename in a specific folder
  * Created new `Filter` object aggregating the functionality of wildcards
  * Created new module `openlane.common.metrics` to aggregate functions dealing with metrics, replacing previous file
    * `__main__` file of this module has three commands, `compare`, `compare-multiple`, `compare-main`:
      * `compare`: Compares two metrics files and generates an output summary
      * `compare-multiple`: Compares two folders of metrics files and generates an output summary
      * `compare-main`: Compares a folder of metric files with metrics fetched from `efabless/openlane-metrics` and generates an output summary
    * New `Metric` object created, encapsulating the metric name, its aggregator, and some other functionality related to evaluating changes in metrics
      * Also allowed some metrics to exclude some groups from aggregation, e.g. `iter` for routing metrics
    * Created "library" of `Metrics` of the various metrics we aggregated so far
      * `aggregate_metrics` now uses the metric library for aggregators as a default value
    * Updated parsing of metrics to more closely match https://github.com/ieee-ceda-datc/datc-rdf-Metrics4ML: the format for modifiers was wrong
* `openlane.state`
   * Created `__main__` file that returns the latest state for a run directory (using `get_latest_file`, with an options to extract the metrics thereof for a file (useful for scripts) 
* `openlane.config.Config`:
  * Updated `refg::` to return the glob string verbatim if the glob returns no results
* `openlane.config.Macro`:
  * Created new method `from_state` that automatically creates a `Macro` object from a `State` object
* Created new object, `ScopedFile`, which is a file that's deleted as soon as it's deconstructed
* `KLayout.*`:
  * Unified generation of commandline arguments using `.get_cli_args()`
  * Updated scripts to use `%OL_METRIC_I` instead of having to use other communication channels
  * Made sure all paths passed to KLayout commands are absolute because KLayout loves to not use CWD for relative paths
* `Magic.*`:
  * Fixed bug when `MAGICRC` is a relative path
* `Netgen.LVS`, `Checker.LVS`:
  * Updated metric names for grammatical consistency 
* Added `httpx` to replace `requests` as the dependency for HTTP functions

## CI and Testing
* Added function in CI to test every PR against the current `main` function
* Added `pytest-xdist` to dev dependencies, so multiple tests can be run in parallel

---

Depends on https://github.com/efabless/openlane2-step-unit-tests/pull/21

### CODE REVIEW

**Comment by donn** [comment]
This is a test message to test if the metric commenter updates the existing comment or makes a new one.

**Comment by donn** [comment]
This is a test message to test if the metric commenter updates the existing comment or makes a new one.

**Comment by donn** [comment]
@openlane-bot wow its like it's a real robot

**Review by kareefardi — COMMENTED** [review]
Couple general additional comments:

1- `design__lint_errors__count` -> `design__lint_error__count` where `error` is without an `s`.
2- I think any change in a critical variable should be reported regardless of value `higher_is_better`. As such, table verbosity `CRITICAL` will always display critical changes.

File: openlane/common/metrics/__main__.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+    All other files are ignored.
+    """
+    path_a, path_b = metric_folders
+    summary, tables = _compare_metric_folders(
+        filter_wildcards, table_verbosity, path_a, path_b
+    )
+    print(summary)
+    if table_out is not None:
+        file = open(table_out, "w", encoding="utf8")
+    print(tables, file=file)
```
Comment (kareefardi) [inline openlane/common/metrics/__main__.py:258]: `file` is undefined if `table_out` is None

File: .github/test_sets/test_sets.yml [hunk 2]
```diff
@@ -2,19 +2,19 @@
...
-    - blink
-    - name: xtea
-      config_file: config.tcl
-    - usb
-    - usb_cdc_core
-    - name: latch
-      config_file: config_good.json
-    - name: latch
-      config_file: config_bad.json
+    # - wbqspiflash
```
[the remark this answers is not in the document]
Comment (donn) [reply .github/test_sets/test_sets.yml:5]: Yes, will remove

File: openlane/common/metrics/__main__.py [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+    session = get_httpx_session(token)
+
+    if commit is None:
+        try:
+            result = session.get(f"https://api.github.com/repos/{repo}/branches/main")
+        except httpx.HTTPStatusError as e:
+            if e.response is not None and e.response.status_code == 404:
+                print(f"main branch of repo {repo} not found.", file=sys.stderr)
+            else:
+                print(f"failed to get info from github API: {404}", file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (donn) [reply openlane/common/metrics/__main__.py:318]: You're correct

File: openlane/common/metrics/metric.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+max_aggregator: MetricAggregator = (-inf, max)
+
+
+@dataclass
+class MetricComparisonResult:
+    """
+    :param metric_name: The name of the metric that has been compared
+    :param gold: The "gold" value being compared against
+    :param new: The new value being evaluated
+    :param delta: Non-``None`` if and only if ``before`` - ``after`` is a valid number.
```
[the remark this answers is not in the document]
Comment (donn) [reply openlane/common/metrics/metric.py:35]: It's rendering just fine here…

File: openlane/common/metrics/util.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        worse = []
+        critical = []
+        remaining = []
+
+        for row in differences:
+            if row.critical is True:
+                critical.append(row)
+            elif row.better is False:
+                worse.append(row)
+            elif (row.delta is not None and row.delta == 0) or row.gold == row.new:
```
[the remark this answers is not in the document]
Comment (donn) [reply openlane/common/metrics/util.py:190]: If it's numeric, `delta`, if not, no, not really

File: openlane/common/toolbox.py [hunk 6]
```diff
@@ -183,19 +183,18 @@ def get_timing_files(
...
 
-            If not specified, the value for `DEFAULT_CORNER` from the SCL will
+            If not specified, the value for ``DEFAULT_CORNER`` from the SCL will
             be used.
         :param prioritize_nl:
             Do not return lib files for macros that have gate-Level Netlists and
             SPEF views.
 
-            If set to ``false``, only lib files are returned.
+            If set to ``false``\\, only lib files are returned.
```
[the remark this answers is not in the document]
Comment (donn) [reply openlane/common/toolbox.py:192]: Double backticks thing

File: openlane/common/metrics/metric.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kareefardi) [reply openlane/common/metrics/metric.py:35]: Yes. Sorry confused this with something else.

File: openlane/steps/openroad.py [hunk 7]
```diff
@@ -283,10 +260,7 @@ def run(self, state_in, **kwargs) -> Tuple[ViewsUpdate, MetricsUpdate]:
                     or_metrics_out[key] = -inf
             metrics_updates.update(or_metrics_out)
 
-        metric_updates_with_aggregates = aggregate_metrics(
-            metrics_updates,
-            timing_metric_aggregation,
-        )
+        metric_updates_with_aggregates = aggregate_metrics(metrics_updates)
```
[the remark this answers is not in the document]
Comment (donn) [reply openlane/steps/openroad.py:263]: Why list metrics twice? This'll just do the aggregation once 'n that's that

File: openlane/common/metrics/__main__.py [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+def compare(
+    metric_files: Tuple[str, str],
+    table_verbosity: TableVerbosity,
+    filter_wildcards: Tuple[str, ...],
+    table_out: Optional[str],
+):
+    """
+    Creates a small summary of the differences between two ``metrics.json`` files.
+    """
+    if table_verbosity == "NONE":
```
[the remark this answers is not in the document]
Comment (donn) [reply openlane/common/metrics/__main__.py:102]: If you just want the highlights and not the table

**Comment by donn** [comment]
@kareefardi I can't say I agree with every change to a "critical" metric being critical. For example-- slack increasing is not "critical" as in demanding attention, but slack decreasing very much is.

What this tells me is that we're not aligned on the definition of critical: my definition of critical was "a change impacting the manufacturability of a design." Any change in I/O pin count, for example, is critical (as evidenced by it having no `higher_is_better` property) as there is no ordinary circumstance in which that would change-- but there are some metrics where only negative changes impact the manufacturability of the design. Meanwhile, OpenROAD `rsz` getting better at its job would increase slacks across all designs and, under this definition, would not add immense noise to the reports without cause for concern.

How did you define critical?

**Comment by donn** [comment]
@kareefardi Bot issue seems to be that the current HEAD of master doesn't have metrics uploaded. Simplest solution would be to just merge this PR.

**Pull request merged**